morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs

```diff
@@ -315,71 +478,463 @@ instance
+  mpretty SubscriptionTraceSubscriptionFailed =
+    "Failed to start all required subscriptions"
+  mpretty (SubscriptionTraceSubscriptionWaitingNewConnection delay) =
+    "Waiting" <+> viaShow delay <+> "before attempting a new connection"
+  mpretty (SubscriptionTraceStart val) =
+    "Starting Subscription Worker, valency" <+> pretty val
+  mpretty (SubscriptionTraceRestart duration desiredVal currentVal) =
+    "Restarting Subscription after"
+      <+> viaShow duration
+      <+> "desired valency"
+      <+> pretty desiredVal
+      <+> "current valency"
+      <+> pretty currentVal
+  mpretty (SubscriptionTraceConnectionExist dst) =
+    "Connection Existed to" <+> mpretty dst
+  mpretty (SubscriptionTraceUnsupportedRemoteAddr dst) =
+    "Unsupported remote target address" <+> mpretty dst
+  mpretty SubscriptionTraceMissingLocalAddress =
+    "Missing local address"
+  mpretty (SubscriptionTraceAllocateSocket addr) =
+    "Allocate socket to" <+> mpretty addr
+  mpretty (SubscriptionTraceCloseSocket addr) =
+    "Closed socket to" <+> mpretty addr
+
+instance MPretty ConnectResult where
+  mpretty ConnectSuccess = "Successful connection"
+  mpretty ConnectSuccessLast = "Successfully connection, reached the valency target. Other ongoing connection attempts will be killed."
+  mpretty ConnectValencyExceeded = "Someone else manged to create the final connection to a target before us."
+
+instance MPretty DnsTrace where
+  -- DnsTrace already has a nice humain readable Show instance
   mpretty = viaShow
 
-instance MPretty (WithIPList (SubscriptionTrace Socket.SockAddr)) where
+instance MPretty Socket.SockAddr where
   mpretty = viaShow
 
-instance (Show peer) => MPretty (WithMuxBearer peer MuxTrace) where
+instance (Show a) => MPretty (NtN.ConnectionId a) where
   mpretty = viaShow
 
+instance MPretty a => MPretty (WithIPList a) where
+  mpretty (WithIPList (LocalAddresses a b c) destAddrs event) =
+    let localAddrs = Trace.trace (show $ catMaybes [a, b, c]) $ catMaybes [a, b, c]
+     in "Subscription(src:"
+          <+> mpretty localAddrs
+          <> ", dest:"
+          <+> mpretty destAddrs
+          <> ", event:"
+          <+> mpretty event
+          <> ")"
+
+instance (MPretty peer) => MPretty (WithMuxBearer peer MuxTrace) where
+  mpretty (WithMuxBearer peer a) = "Mux bearer" <+> mpretty peer <+> "with:" <+> viaShow a
+
+instance
+  ( forall (st :: ps). Show (ClientHasAgency st),
+    forall (st :: ps). Show (ServerHasAgency st),
+    forall (st :: ps) (st' :: ps). Show (Message ps st st'),
+    MPretty peer
+  ) =>
+  MPretty (WithMuxBearer peer (NtN.TraceSendRecv ps))
+  where
+  mpretty (WithMuxBearer peer a) = "Mux bearer" <+> mpretty peer <+> "with:" <+> mpretty a
+
 instance MPretty a => MPretty [a] where
-  mpretty xs = list $ map mpretty xs
+  -- mpretty xs = list $ map mpretty xs
+  mpretty xs =
+    "["
+      <> concatWith (\d1 d2 -> d1 <> "," <+> d2) (map mpretty xs)
+      <> "]"
```

Comment:
Check out how prettyprinter's [`prettyList`](https://hackage.haskell.org/package/prettyprinter-1.7.0/docs/Prettyprinter.html#v:prettyList) is implemented. We can do the same here as a simplification

Reply:
@Infinisil The problem with `prettyList`'s implementation is that it uses `list`. That function will automatically create line breaks for every element of the list when the line has more than `n` characters (default is 80). Example of an output:

```
[1
, 2
, 3]
```

Of course, one solution would be to set the number of characters to a much higher value. 

What do you think?

Reply:
@Infinisil  `prettyList` uses the `list` function which might introduce line breaks. However, I just found the `hsep` function which doesn't